dojo_plugin/pages/canvas.py

```diff
@@ -0,0 +1,225 @@
+import json 
+import logging 
+import requests
+import time 
+
+from flask import request, Blueprint, abort
+from CTFd.models import Users
+from CTFd.utils.decorators import authed_only
+
+from ..models import Dojos, DojoChallenges, DojoStudents, DojoStudents
+from .course import grade
+from ..utils.dojo import dojo_route
+
+canvas = Blueprint("canvas", __name__)
+
+log=logging.getLogger(__name__)
+
+""" Endpoint to do a full class canvas sync """
+@canvas.route("/dojo/<dojo>/admin/canvas_sync")
+@dojo_route
+@authed_only
+def canvas_sync_all(dojo):
+    if not dojo.course:
+        abort(404)
+
+    if not dojo.is_admin():
+        abort(403)
+    
+    posting_results = do_canvas_sync(dojo)
+        
+    return json.dumps(posting_results, indent=2)
+
+
+"""
+This is called from the the DojoFlag class on a successful flag submission.
+It updates a single user's Canvas grade for the module that the the user just sucessfully submitted 
+"""
+@authed_only
+def sync_challenge_to_canvas(challenge_id, user_id):
+    dojo_chal = DojoChallenges.query.filter(DojoChallenges.challenge_id == challenge_id).first()
+    if dojo_chal:
+        dojo = Dojos.query.filter(Dojos.dojo_id == dojo_chal.dojo_id).first()
+
+        posting_results = do_canvas_sync(dojo, user_id=user_id, module_id=dojo_chal.module.id)
+    
+        log.info("Result from post", json.dumps(posting_results, indent=2))
+
+
+""" 
+This function will retrieve the grade information, format it, and send to canvas
+However, if both user_id and module_id are supplied it will do a single user/module update 
+"""
+def do_canvas_sync(dojo, user_id=None, module_id=None):
+    
+    canvas_token = dojo.course.get("canvas_token","")
+    
+    canvas_api_host = dojo.course.get("canvas_api_host","")
+    canvas_course_id = dojo.course.get("canvas_course_id",0)
+    if len(canvas_token) == 0:
+        return {"status": "completed", "message": "Missing canvas_token in course.xml"}
```

Comment:
What is a course.xml? I thought we used yaml?

Reply:
It's a mistype, I'll fix.